website/blog/2022-11-31-simple-2.mdx

```diff
@@ -0,0 +1,115 @@
+---
+slug: simple-functions-2
+title: "Simple Functions: Efficient Complex Types"
+authors: [lsakka]
+tags: [tech-blog, simple-functions]
+---
+_This blogpost is part of a series of blog posts that discuss different features and optimizations of the simple function interface_.
+
+
+## Efficient Complex Types
+In this blogpost, we will discuss two major changes that we did recently to the simple function interface to make its performance comparable to the vector function implementations for functions that produce or consume complex types (Arrays, Maps and Rows).
+
+To show how much simpler simple functions are. The figure below shows a function _NestedMapSum_ that is written in both the simple and vector interfaces. The function consumes a nested map and computes the summations of all values and keys.
+Note that the vector function implementation is minimal, without any special optimization. Adding optimizations will make it even longer.
+
+<figure>
+ <img src="/img/simple1_2.png"/>
+ <figcaption> NestedMapSum function implemented using vector(left) and simple(right) interfaces.</figcaption>
+</figure>
+
+
+### View types for inputs
+The previous representations of input complex types in the simple function interface were extremely expensive. Data from vectors used to be copied into std containers and passed to simple functions to process it.
+
+For example, to execute the function _array_first_ that returns the first element in an array, the SimpleFunctionAdapter used to read the input array from the input vector for every row and construct an _std::vector<std::optional<\>\>_, (std::optional is used to express the nullity) and pass it to the function for execution. The graph below illustrates the previous implementation approach.
```

Comment:
As a high-level, the focus could be a bit more on the positive side of the new API, rather than describing in too much detail how the old was ineficient. Readers would mostly be interested in learning about the new API anyway. 

Condensing this in maybe a phrase or two would also make the post a bit shorter and more concise, which is good.

Reply:
make sense changed